src/key_io.cpp

```diff
@@ -78,82 +78,128 @@ class DestinationEncoder
 
 CTxDestination DecodeDestination(const std::string& str, const CChainParams& params, std::string& error_str, std::vector<int>* error_locations)
 {
-    std::vector<unsigned char> data;
     uint160 hash;
     error_str = "";
 
-    // Note this will be false if it is a valid Bech32 address for a different network
-    bool is_bech32 = (ToLower(str.substr(0, params.Bech32HRP().size())) == params.Bech32HRP());
-
-    if (!is_bech32 && DecodeBase58Check(str, data, 21)) {
+    struct decodeState {
+        const bech32::DecodeResult bech32DecodeResult;
+        uint8_t maxBase58CheckChars = 21;
+        uint8_t maxBase58Chars = 100;
+        bool is_base58 = false;
+        bool is_base58Check = false;
+        bool is_bech32 = false;
+        bool is_validBech32Chars = false;
+        std::pair<std::string, std::vector<int>> bech32DecodeErrors;
+        std::vector<unsigned char> base58DataRaw, base58DataCheck, bech32Data;
+        std::string networkLabel;
+        // Perform base58/bech32 decoding on the input string
+        decodeState(std::string str, std::string chainName) : bech32DecodeResult(bech32::Decode(str))
+        {
+            is_base58 = DecodeBase58(str, base58DataRaw, maxBase58Chars);
+            is_base58Check = DecodeBase58Check(str, base58DataCheck, maxBase58CheckChars);
+            is_bech32 = bech32DecodeResult.encoding != bech32::Encoding::INVALID;
+            networkLabel = (chainName == "main" || chainName == "test") ? chainName + "net" : chainName;
```

Comment:
Bitcoin shouldn't be called "mainnet" in user-facing errors.

Reply:
This has been resolved by adding a function to chaintype.cpp called ChainTypeToDisplayString. This decodes the ChainType Enum into user-facing strings. Including using 'Bitcoin' to reference mainnet

Impoper chain errors are now displayed  for example 
```"error": "Invalid chain prefix for Bitcoin. Expected bc got tb"```